### PR TITLE
[WIP] Aleph learner

### PR DESCRIPTION
**Pull request by QuintenBruynseraede**
Notable changes:
- Added an abstract class for Learners, with provides a learn() method
- Structured code to collect Learners in one folder
- Added evaluation functions use when evaluating newly created clauses
- Implemented Aleph learner

### CODE REVIEW

**Comment by QuintenBruynseraede** [comment]
Update 22/11:
The basic Aleph search algorithm has been implemented and works for a simple example (see ```tests/aleph_tests.py```). However, Aleph uses two ways of generating new clauses: 
- add a new atom to the end
- instantiate a variable

Instantiation of variables is not yet supported in ```TopDownHypothesisSpace```, so that limits the current hypothesis space to clauses that only contain, i.e. 
``` child_of_homer(X) :- father(homer,X). ``` is not learnable yet.

File: loreleai/learning/hypothesis_space.py [hunk 1]
```diff
@@ -156,8 +160,14 @@ def initialise(self) -> None:
...
 
-            # create empty clause
-            clause = Body()
+            # create empty clause or use initial clause
+            if initial_clause:
+                clause = initial_clause if isinstance(initial_clause,Body) else initial_clause.get_body() 
+            else:
+                clause = Body()
+            if len(clause.get_literals()) > 0 and len(clause.get_variables()) < self._head_constructor.get_arity():
+                raise AssertionError("Cannot provide an initial clause with fewer distinct variables than the head predicate!")
```
Comment (sebdumancic) [inline loreleai/learning/hypothesis_space.py:169]: this could be made possible - additional variables will appear after extension

**Comment by sebdumancic** [comment]
this is very good.
A few suggestions:
 - I would make bottom clause generation/saturation a separate function so that it is easy to use in other implementations
 - re-use the Hypothesis space between iterations. You could cache the evaluations in the previous iterations and potentially save significant time

Otherwise, it looks good!
I'll merge it

**Comment by QuintenBruynseraede** [comment]
Thanks for the review. I will try to find some time to implement the suggestion you mentioned (in particular, caching evaluations is interesting imo)!

**Comment by sebdumancic** [comment]
I have just added support for that.
Hypothesis space now has three new methods:
 - add_to_cache(clause, key, val) to add key-val pair to the cache
 - retrieve_from_cache(clause, key) to retrieve the values associated with the key
 - remove_from_cache(clause, key) to remove the key from the cache

**Comment by QuintenBruynseraede** [comment]
New addition: a LearnResult class (see abstract_learners.py) that holds statistics about the learning process (things like number of iterations, number of queries etc.)

Caching now works across iterations too.

**Pull request merged**